#### Metrics Calculation for specific genes
1. Upload a list of your genes (as csv, txt, tsv file) or write them in the text field.
2. Choose your genome release and CADD version and then click on the “Generate metrics” button.
3. Now all the metrics will load in one line graph. (If you want to see one metric, double click on the name on the legend. If you want to see more than one metrics, deselect all others b clicking once on the name on the legend.)
- If you want to know which variants were used for calculating, together with their annotations, you can look at the table. You may choose if you want to look at the ClinVar or CADD annotations or both.
For ClinVar only these annotations were kept: 'AlleleID', 'Type_x', 'Name', 'GeneID_x', 'GeneSymbol', 'Origin', 'OriginSimple', 'Chromosome', 'ReviewStatus', 'NumberSubmitters', 'VariationID', 'PositionVCF', 'ReferenceAlleleVCF', 'AlternateAlleleVCF', 'ClinicalSignificance'
- To see how many variants were used per gene and if they are pathogenic or benign you can look at the bar chart (it might not be visible if you used a lot of variants, you could still zoom in). Below the bar chart is also a table that summarizes the information from the bar chart.
